mitodl/open-discussions
# Open discussions Elastic/Opensearch performance updates

Added a configuration setting for specifying the number of index replicas to create when making a new elasticsearch index.

#### Pre-Flight checklist

#### What are the relevant tickets?
https://github.com/mitodl/ol-infrastructure/issues/1199

#### What's this PR do?
This PR will add a configuration setting that allows us to specify the number of index replicas when the application creates a new index in OpenSearch. Our focus is on search performance rather than ingest performance. For improved search performance we should have a replica of all the data on each node in the cluster, which will allow any node that receives a search request to service that request rather than forwarding it to another node 33% of the time (in the case of a 3 node cluster with 1 replica, the current config). 

In confusing ES terminology "1 replica" == 2 copies of the data, a primary + replica. So, "2 replicas" == 3 copies of the data, a primary and 2 replicas. 🤷 

We set the default replica count to 2, which is appropriate for a 3 node cluster. DevStack implementations will not break for this, the cluster will simply be unable to allocate the additional replicas (which is already what happens because the default setting is 1 -> 2 copies of the data but only one node). 

Additionally, we are going to set the refresh interval on new indices to 60 seconds. The default value is 1 second. This change will prevent OS/ES from doing constant "refreshes" during our reindex process that are of no value to us with our given usage pattern. We execute all of our searches against the alias for an index, rather than the explicit index name. Once our reindex process is completed and the new index is fully populated, we swap the alias over and being executing searches against the new index. A "refresh" updates an index to a usable state for search with all the new data that has been added since the last "refresh" took place. Because never execute a search against an index that is being actively updated, there is no value provided by doing this every second. 
[
More information on refreshes.](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-refresh.html)

Additionally-Addtionally, we're going to add an ELASTICSEARCH_CONNECTIONS_PER_NODE setting to allow us to override the default size of the connection pool(s) created by ElasticSearch-py in urllib3. The default set by elasticsearch-py is 10, and that is our default as well, but we have an option to adjust it per environment based on performance issues observed. 

#### How should this be manually tested?
There isn't much point to testing this in a devstack. We will want to check on the CI/RC environment to confirm that new indices are created with the appropriate setting populated. It might not be possible to asses the performance impact of these changes in CI/RC though, given then have low search load. 

#### Where should the reviewer start?
(Optional)

#### Any background context you want to provide?
(Optional)

#### Screenshots (if appropriate)
(Optional)

#### What GIF best describes this PR or how it makes you feel?
(Optional)
